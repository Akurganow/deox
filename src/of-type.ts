import { OperatorFunction } from 'rxjs'
import { filter } from 'rxjs/operators'

import { ActionCreator } from './create-action-creator'
import { Action, AnyAction } from './create-action'
import { getType } from './get-type'
import { castArray } from './utils'

/**
 * Filter actions emitted by the source Observable by only emitting those that
 * are compatible with specified action(s) or action creator(s) or action type(s).
 *
 * @example
 * action$.pipe(
 *   ofType(foo),
 *   ...
 * )
 * @example
 * action$.pipe(
 *   ofType([foo, bar]),
 *   ...
 * )
 */
export function ofType<
  TSource extends AnyAction,
  TActionCreator extends ActionCreator<TSource>,
>(
  actionCreators: TActionCreator | ReadonlyArray<TActionCreator>
): OperatorFunction<TSource, ReturnType<TActionCreator>>
export function ofType<TSource extends AnyAction, TAction extends TSource>(
  actions: TAction | ReadonlyArray<TAction>
): OperatorFunction<TSource, TAction>
export function ofType<
  TSource extends AnyAction,
  TType extends TSource['type']
>(
  types: TType | ReadonlyArray<TType>
): OperatorFunction<
  TSource,
  TSource extends Action<infer T> ? (T extends TType ? TSource : never) : never
>
export function ofType(keys: ActionCreator<AnyAction> | AnyAction | string) {
  const types = castArray(keys).map(key =>
    typeof key === 'string' ? key : getType(key)
  )

  return filter<AnyAction>(action => types.includes(action.type))
}
