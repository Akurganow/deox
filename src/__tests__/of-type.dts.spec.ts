import { of } from 'rxjs'

import { createAction } from '../create-action'
import { ofType } from '../of-type'

// @dts-jest:group ofType

const a = createAction('a')
const b = createAction('b')
const c = createAction('c')

// @dts-jest:pass:snap
of(a(), b(), c()).pipe(ofType(a))

// @dts-jest:pass:snap
of(a(), b(), c()).pipe(ofType([a, b]))

// @dts-jest:pass:snap
of(a(), b(), c()).pipe(ofType(a()))

// @dts-jest:pass:snap
of(a(), b(), c()).pipe(ofType([a(), b()]))

// @dts-jest:pass:snap
of(a(), b(), c()).pipe(ofType('a'))

// @dts-jest:pass:snap
of(a(), b(), c()).pipe(ofType(['a', 'b']))
