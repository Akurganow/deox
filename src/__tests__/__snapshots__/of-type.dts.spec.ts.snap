// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ofType of(a(), b(), c()).pipe(ofType('a')) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; }>"`;

exports[`ofType of(a(), b(), c()).pipe(ofType(<const>['a', 'b'])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; }>"`;

exports[`ofType of(a(), b(), c()).pipe(ofType([a(), b()])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; }>"`;

exports[`ofType of(a(), b(), c()).pipe(ofType([a, b])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; }>"`;

exports[`ofType of(a(), b(), c()).pipe(ofType(a())) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; }>"`;

exports[`ofType of(a(), b(), c()).pipe(ofType(a)) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; }>"`;

exports[`ofType of(a(), b(), c(), d()).pipe(ofType(<const>['a', b(), c])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; } | { type: \\"c\\"; }>"`;

exports[`ofType of<AnyAction>().pipe(ofType('a')) (type) should match snapshot 1`] = `"Observable<{ type: string; } & { type: \\"a\\"; }>"`;

exports[`ofType of<AnyAction>().pipe(ofType(<const>['a', b(), c])) (type) should match snapshot 1`] = `"Observable<{ type: \\"b\\"; } | ({ type: string; } & { type: \\"a\\"; }) | { type: \\"c\\"; }>"`;

exports[`ofType of<AnyAction>().pipe(ofType(['a', 'b'])) (type) should match snapshot 1`] = `"Observable<({ type: string; } & { type: \\"a\\"; }) | ({ type: string; } & { type: \\"b\\"; })>"`;

exports[`ofType of<AnyAction>().pipe(ofType([a(), b()])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; }>"`;

exports[`ofType of<AnyAction>().pipe(ofType([a, b])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; }>"`;

exports[`ofType of<AnyAction>().pipe(ofType(a())) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; }>"`;

exports[`ofType of<AnyAction>().pipe(ofType(a)) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; }>"`;

exports[`ofType of<Redux.Action>().pipe(ofType('a')) (type) should match snapshot 1`] = `"Observable<Action<any> & { type: \\"a\\"; }>"`;

exports[`ofType of<Redux.Action>().pipe(ofType(<const>['a', b(), c])) (type) should match snapshot 1`] = `"Observable<{ type: \\"b\\"; } | { type: \\"c\\"; } | (Action<any> & { type: \\"a\\"; })>"`;

exports[`ofType of<Redux.Action>().pipe(ofType(['a', 'b'])) (type) should match snapshot 1`] = `"Observable<(Action<any> & { type: \\"a\\"; }) | (Action<any> & { type: \\"b\\"; })>"`;

exports[`ofType of<Redux.Action>().pipe(ofType([a(), b()])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; }>"`;

exports[`ofType of<Redux.Action>().pipe(ofType([a, b])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; }>"`;

exports[`ofType of<Redux.Action>().pipe(ofType(a())) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; }>"`;

exports[`ofType of<Redux.Action>().pipe(ofType(a)) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; }>"`;

exports[`ofType of<any>().pipe(ofType('a')) (type) should match snapshot 1`] = `"Observable<any>"`;

exports[`ofType of<any>().pipe(ofType(<const>['a', b(), c])) (type) should match snapshot 1`] = `"Observable<any>"`;

exports[`ofType of<any>().pipe(ofType(['a', 'b'])) (type) should match snapshot 1`] = `"Observable<any>"`;

exports[`ofType of<any>().pipe(ofType([a(), b()])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; }>"`;

exports[`ofType of<any>().pipe(ofType([a, b])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; }>"`;

exports[`ofType of<any>().pipe(ofType(a())) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; }>"`;

exports[`ofType of<any>().pipe(ofType(a)) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; }>"`;
